### PR TITLE
[DDW-149] Undelegating a wallet

### PR DESCRIPTION
**Pull request by yakovkaravelov**
This PR implemented `Undelegate wallet` dialog on wallet settings screen
[Jira Ticket](https://jira.iohk.io/browse/DDW-149)

## Todos
- [x] Implement `Undelegate wallet` dialog on wallet settings screen
- [x] On the "undelegation wizard" rename "Deposits" to "Deposits reclaimed" and show the value in green.

## Testing Checklist
- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1612769819154400)

### App
- [x] Go to `Wallet` => `Settings` and click on `Undelegate Wallet` button to open `Undelegate Wallet` confirmation dialog and make sure `Undelegate Wallet` confirmation dialog matches with wireframe
- [x] Tick checkboxes and fill in `spending password` to confirm and make sure `Undelegate Wallet` result dialog matches with wireframe

### Test Cases
```Gherkin
Scenario 1 - Delegating a wallet options visible in wallet settings when wallet Undelegated
Given that I have a wallet fully synced in Daedalus
And that wallet has not already been delegated
When I navigate to the wallet settings screen
Then I will see the option to "Delegate your wallet" 
And I will see a button that says "Delegate"

Scenario 2 - Undelegating a wallet options visible in wallet settings when wallet Delegated
Given that I have a wallet fully synced in Daedalus
And that wallet has already been delegated
When I navigate to the wallet settings screen
Then I will see the option "Undelegating your wallet" 
And I will see a button that says "Undelegate"

Scenario 3 - Verify Undelegation wizard available
Given that I have a wallet that has already been delegated
And I navigate to the wallet settings screen
When I click on the "undelegate" button
Then the undelegation wizard should open

Scenario 4 - Verify Delegation wizard available
Given that I have a wallet that has not already been delegated
And I navigate to the wallet settings screen
When I click on the "delegate" button
Then the Delegation wizard should open

Scenario 5 - Verify Undelegation wizard success
Given that I am on the undelegation wizard
And I follow the instructions
When I complete the wizard
Then I should see an undelegation success message

Scenario 6 - Verify Delegation wizard success
Given that I am on the Delegation wizard
And I follow the instructions
When I complete the wizard
Then I should see a Delegation success message

Scenario 7 - Verify Undelegation has been successful
Given that I have completed the undelegation wizard
And I have navigated to the Stake pools screen, delegation centre tab
Then the wallet should be shown as undelegated
And the wallet should have received back the 2ADA Delegation deposit
And I should have the option to delegate wallet
And the delegation change will take place in +2 epochs

Scenario 8 - Verify Delegation has been successful
Given that I have completed the Delegation wizard
And I have navigated to the Stake pools screen, delegation centre tab
Then the wallet should be shown as Delegated
And I should have the option to Undelegate wallet
And the delegation change will take place in 2+epochs

Scenario 9 - Verify Delegate/Undelegate unavailable for a Byron Wallet
Given that I have a Byron Wallet fully synced in Daedalus
WhenI navigate to the wallet settings screen
Then I should not see an option to Delegate/Undelegate

Scenario 10 - Delegate/Undelegate unavailable while wallet is still syncing
Given that I have added a wallet to Daedalus
And it is still in loading state
When when I navigate to wallet settings
Then I see that the Delegation/Undelegation button is inactive
And I see a message explaining that this wallet is currently synchronizing with the blockchain
```

### English
image
image
image
image
image

### Japanese
image
image
image
image
image

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes and shows some screenshots or animated GIFs of important UI changes
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

**Review by nikolaglumac — CHANGES_REQUESTED** [review]
@yakovkaravelov here's what I have found so far:

1. When clicking on "Undelegate" button, Daedalus breaks and fallbacks to white-screen with following errors in the console:

Screenshot 2021-02-08 at 09 31 53

2. The bottom margin is too big - it should match the one from the element above the "Undelegation" box:

margin

3. "Undelegation" box should be shown **only** for delegated wallets and never for Byron wallets:

Screenshot 2021-02-08 at 09 31 20

4. We need to decide what to do while the wallet is restoring. Please start a thread on Slack and ask @darko-mijic and @a-rukin for guidance.

Thanks!

**Comment by yakovkaravelov** [comment]
@nikolaglumac @a-rukin @gnpf @mioriohk Please review/test. Thanks.
FYI: Will add screenshots soon when I have some funds in my wallet.

**Comment by yakovkaravelov** [comment]
@nikolaglumac @a-rukin @gnpf @mioriohk Please review/test. Thanks.

**Comment by yakovkaravelov** [comment]
@nikolaglumac @daniloprates @gnpf @mioriohk @ManusMcCole Please review/test, thanks.

File: source/renderer/app/components/staking/delegation-setup-wizard/DelegationStepsConfirmationDialog.js
Comment (nikolaglumac) [inline source/renderer/app/components/staking/delegation-setup-wizard/DelegationStepsConfirmationDialog.js:304]: @yakovkaravelov do we still nee dthis hack=

File: source/renderer/app/components/wallet/settings/UndelegateWalletConfirmationDialog.scss
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/settings/UndelegateWalletConfirmationDialog.scss:88]: @yakovkaravelov please revert to `isSubmitting`

File: source/renderer/app/components/wallet/settings/WalletSettings.js
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/settings/WalletSettings.js:11]: @yakovkaravelov please rename both of these variables to:
```
IS_WALLET_PUBLIC_KEY_SHARING_ENABLED
IS_WALLET_UNDELEGATION_ENABLED
```

File: source/renderer/app/api/staking/types.js
Comment (nikolaglumac) [inline source/renderer/app/api/staking/types.js:104]: @yakovkaravelov is this used anywhere at all?

**Review by nikolaglumac — CHANGES_REQUESTED** [review]
@yakovkaravelov please see my comments. I will do another review tomorrow.

**Comment by nikolaglumac** [comment]
@tomothespian please review the part with HW interaction :pra

**Comment by tomothespian** [comment]
> @tomothespian please review the part with HW interaction 🙏

@nikolaglumac no problem

File: source/renderer/app/components/wallet/settings/DelegateWalletButton.js
Comment (tomothespian) [inline source/renderer/app/components/wallet/settings/DelegateWalletButton.js:32]: @yakovkaravelov maybe better function naming would be `onDelegate`

File: source/renderer/app/components/wallet/settings/DelegateWalletButton.scss
Comment (tomothespian) [inline source/renderer/app/components/wallet/settings/DelegateWalletButton.scss:8]: @yakovkaravelov can you wrap component and use global to override this instead of `!important`
e.g.

```
.componentWrapper {
  :global {
      .Button_root {...}
  }
}
```

File: source/renderer/app/components/wallet/settings/UndelegateWalletButton.js
Comment (tomothespian) [inline source/renderer/app/components/wallet/settings/UndelegateWalletButton.js:33]: Same as above. Maybe `onUndelegate`

File: source/renderer/app/components/wallet/settings/UndelegateWalletButton.scss
Comment (tomothespian) [inline source/renderer/app/components/wallet/settings/UndelegateWalletButton.scss:9]: Same as above. Use wrapper + `:global` instead of !important

File: source/renderer/app/components/wallet/settings/UndelegateWalletConfirmationResultDialog.js
Comment (tomothespian) [inline source/renderer/app/components/wallet/settings/UndelegateWalletConfirmationResultDialog.js:43]: @yakovkaravelov can we remove "Result" from class name

File: source/renderer/app/components/wallet/settings/WalletSettings.js
Comment (tomothespian) [inline source/renderer/app/components/wallet/settings/WalletSettings.js:255]: This can be one `if` clause

**Comment by a-rukin** [comment]
@yakovkaravelov 
for delegated wallet I can't see new block + paddings are mad 
image

File: source/renderer/app/containers/wallet/dialogs/settings/UndelegateWalletDialogContainer.js
Comment (tomothespian) [inline source/renderer/app/containers/wallet/dialogs/settings/UndelegateWalletDialogContainer.js:14]: @yakovkaravelov I'm not sure that this `any` declaration is ok. We have store types declared e.g. `InjectedDialogContainerProps` and similar.

File: storybook/stories/staking/Undelegate.stories.js
Comment (tomothespian) [inline storybook/stories/staking/Undelegate.stories.js:13]: @yakovkaravelov can you also add a story for hardware wallets. Use knobs to switch `isHardwareWallet` true / false and generate one more wallet with `isHardwareWallet` flag set to `true`

**Review by tomothespian — CHANGES_REQUESTED** [review]
@yakovkaravelov good job. I left just a few comments. Can you please take a look?

**Comment by tomothespian** [comment]
@yakovkaravelov I don't see DELEGATE / UNDELEGATE in wallet settings for Ledger Nano X. For Nano S is presented as well as for Trezor

Screen Shot 2021-02-12 at 10 34 02

**Comment by tomothespian** [comment]
@yakovkaravelov negative fees when you want to undelegate  Hardware wallet! 

Screen Shot 2021-02-12 at 10 36 53

**Comment by tomothespian** [comment]
@yakovkaravelov Maybe we should trigger Hardware wallet Connecting -> Confirming (Undelegate) once Checkboxes are checked?
cc @nikolaglumac @darko-mijic

**Comment by tomothespian** [comment]
I can't undelegate hardware wallet. But this is probably because of negative fees. Can you please check that?
Screen Shot 2021-02-12 at 10 43 50

File: source/renderer/app/components/wallet/settings/DelegateWalletButton.scss
Comment (yakovkaravelov) [reply source/renderer/app/components/wallet/settings/DelegateWalletButton.scss:8]: @tomothespian Good catch. I didn't make this one new in this PR. Just copied from other existing style, and didn't try to beautify it. Will fix now.

File: source/renderer/app/components/wallet/settings/UndelegateWalletConfirmationResultDialog.js
Comment (yakovkaravelov) [reply source/renderer/app/components/wallet/settings/UndelegateWalletConfirmationResultDialog.js:43]: @tomothespian There is already other component which is named `UndelegateWalletConfirmationDialog`.
This one is for result.

File: source/renderer/app/components/wallet/settings/UndelegateWalletButton.scss
Comment (yakovkaravelov) [reply source/renderer/app/components/wallet/settings/UndelegateWalletButton.scss:9]: @tomothespian Good catch. I didn't make this one new in this PR. Just copied from other existing style, and didn't try to beautify it. Will fix now.

File: source/renderer/app/components/wallet/settings/UndelegateWalletButton.js
Comment (yakovkaravelov) [reply source/renderer/app/components/wallet/settings/UndelegateWalletButton.js:33]: 👍

File: source/renderer/app/components/wallet/settings/DelegateWalletButton.js
Comment (yakovkaravelov) [reply source/renderer/app/components/wallet/settings/DelegateWalletButton.js:32]: 👍

File: source/renderer/app/components/wallet/settings/WalletSettings.js
Comment (yakovkaravelov) [reply source/renderer/app/components/wallet/settings/WalletSettings.js:255]: @tomothespian I did this intentionally because we will remove config variable eventually.
Wanted to avoid editing valid if clause in 253 line when removing config.

File: source/renderer/app/containers/wallet/dialogs/settings/UndelegateWalletDialogContainer.js
Comment (yakovkaravelov) [reply source/renderer/app/containers/wallet/dialogs/settings/UndelegateWalletDialogContainer.js:14]: 👍

File: storybook/stories/staking/Undelegate.stories.js
Comment (yakovkaravelov) [reply storybook/stories/staking/Undelegate.stories.js:13]: 👍

File: source/renderer/app/components/wallet/settings/WalletSettings.js
Comment (tomothespian) [reply source/renderer/app/components/wallet/settings/WalletSettings.js:255]: @yakovkaravelov aha I see. Then we are good with this one. Resolved!

**Review by a-rukin — APPROVED** [review]
LGTM

**Comment by ManusMcCole** [comment]
Hi Yakov. Tested on build 16479 Testnet

For scenario 10. If the wallet is in loading state and delegation state is unknown it says the wallet is not delegated. Im not seeing the expected text "This wallet is currently synchronizing with the blockchain, so its delegation status is currently unknown, and undelegation is not possible."See <a href="https://drive.google.com/file/d/1j1qg8jAnDnV6lsxaNyVg47EjCPBBYHWd/view?usp=sharing">screenshot</a>

For hardware wallets Undelegation I am still seeing the negative fees issue @tomothespian reported earlier. See <a href="https://drive.google.com/file/d/1Ep1PoCOM2U8FLOWZPqWEdcb56XVWIwQW/view?usp=sharing
">screenshot</a>

**Comment by tomothespian** [comment]
@topgunpro I found 2 issues and they are probably related to the comment I left you on Friday:
https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1613146075237100?thread_ts=1612769819.154400&cid=GGKFXSKC6

1. When I try to undelegate hardware wallet a negative fee is presented

Screen Shot 2021-02-15 at 10 27 26

2. When I want to delegate hardware wallet for the first time deposit is also included in the fee and presented separately as deposit as well

Screen Shot 2021-02-15 at 10 28 36

**Comment by nikolaglumac** [comment]
@topgunpro I have a delegated wallet "Shelley wallet" and I still get the "Delegate" button and not the "Undelegate" on it's settings screen...

Screenshot 2021-02-15 at 14 31 36
Screenshot 2021-02-15 at 14 31 40

**Comment by yakovkaravelov** [comment]
@nikolaglumac @daniloprates @tomislavhoracek @ManusMcCole @gabriela-ponce Please review/test. Thanks.

**Comment by ManusMcCole** [comment]
Hi @yakov.Here is the QA report. Tested on Mainnet build 16585. All tests passed except for scenario 10 and the rewards wallet scenario thats stll waiting for the API.

Scenario 10.The message "This wallet is currently synchronizing with the blockchain, so its delegation status is currently unknown, and undelegation is not possible" does not seem to be visible for wallets that are in loading state.When the wallet reached 94.5% loaded the corrrect message appeared briefly and then disappeared again.See <a href="https://drive.google.com/file/d/1pM5TZ0iwQ8Th6OOd7VDQP5GT1FoFtlH0/view?usp=sharing
">video</a> and <a href="https://drive.google.com/file/d/1ut0msPlIkEZ8BUEUPtDmJ16fKe_JtnQX/view?usp=sharing
">video</a>

**Comment by ManusMcCole** [comment]
Hi @yakovkaravelov. Previous scenarios are all fixed.Tested on build 16602 Mainnet

For Scenario 7 when a wallet is undelegated. The changes are not showing up in the Wallets screen after. See <a href="https://drive.google.com/file/d/1wY5Uk7JEojLgR7yw-ufzyTCcqmj27tOr/view?usp=sharing">video</a>

**Comment by yakovkaravelov** [comment]
> Hi @yakovkaravelov. Previous scenarios are all fixed.Tested on build 16602 Mainnet
> 
> For Scenario 7 when a wallet is undelegated. The changes are not showing up in the Wallets screen after. See [video](https://drive.google.com/file/d/1wY5Uk7JEojLgR7yw-ufzyTCcqmj27tOr/view?usp=sharing)

That's because you tested before undelegation transaction was finished.
Please check after that transaction is finished. @ManusMcCole

**Review by ManusMcCole — APPROVED** [review]
Looks good

**Comment by ManusMcCole** [comment]
> > Hi @yakovkaravelov. Previous scenarios are all fixed.Tested on build 16602 Mainnet
> > For Scenario 7 when a wallet is undelegated. The changes are not showing up in the Wallets screen after. See [video](https://drive.google.com/file/d/1wY5Uk7JEojLgR7yw-ufzyTCcqmj27tOr/view?usp=sharing)
> 
> That's because you tested before undelegation transaction was finished.
> Please check after that transaction is finished. @ManusMcCole

As we discussed on the call it seems that issue was caused by having a hardware wallet plugged in. I will investigate this further and create a Ticket for it

**Review by tomislavhoracek — COMMENTED** [review]
@yakovkaravelov please merge the latest develop branch and fix conflicts!

**Comment by yakovkaravelov** [comment]
> @yakovkaravelov please merge the latest develop branch and fix conflicts!

@tomislavhoracek @nikolaglumac Latest develop was merged. Thanks.

**Comment by nikolaglumac** [comment]
@yakovkaravelov I still see the same issue. My wallet "Shelley wallet" is delegated but I still get to see the "Delegate your wallet" instead of "Undelegate your wallet" box on the "Settings screen" 😢 

Screenshot 2021-02-19 at 09 39 07
Screenshot 2021-02-19 at 09 39 12

**Comment by yakovkaravelov** [comment]
@nikolaglumac That's because something happened on your machine as well.
Please unplug hardware wallet, yubi key, and create new wallet and try again.

**Comment by nikolaglumac** [comment]
> @nikolaglumac That's because something happened on your machine as well.
> Please unplug hardware wallet, yubi key, and create new wallet and try again. (Don't use that wallet `Shelley wallet` anymore. It is already broken wallet)

I don't have one connected at all.

**Comment by nikolaglumac** [comment]
There should be no bottom margin here @yakovkaravelov :
Screenshot 2021-02-19 at 10 12 50
cc @alexander-rukin

**Comment by alexander-rukin** [comment]
> There should be no bottom margin here @yakovkaravelov :
> Screenshot 2021-02-19 at 10 12 50
> cc @alexander-rukin

yes, please remove 🙏

**Comment by nikolaglumac** [comment]
@yakovkaravelov please check the Flow errors too 🙏 
(these are caused by the recent merge!)

Screenshot 2021-02-19 at 10 25 27

File: source/renderer/app/api/api.js
Comment (nikolaglumac) [inline source/renderer/app/api/api.js:1003]: @yakovkaravelov please replace with `depositsArray`

File: source/renderer/app/config/walletsConfig.js
Comment (nikolaglumac) [inline source/renderer/app/config/walletsConfig.js:43]: @yakovkaravelov must not forget to change this to `false` before merging the PR.

**Comment by ManusMcCole** [comment]
Hi @yakovkaravelov

The issue we were seeing with software wallets not updating to reflect the recently changed status is also affecting Hardware Wallets now. See <a href="https://drive.google.com/file/d/126WRZNsUrZbC0SaqDI_ZGYg3SIw5peMe/view?usp=sharing">video</a>

File: source/renderer/app/components/wallet/settings/WalletSettings.js
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/settings/WalletSettings.js:255]: @yakovkaravelov this is not good. We need to track active status too. Thank you Yakov!

**Review by miorsufianiohk — APPROVED** [review]
LGTM. Only tested undelegated wallet without reward and delegated wallet without reward (as instructed) on build 16647. Well done @yakovkaravelov 👍

File: source/renderer/app/components/wallet/settings/WalletSettings.js
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/settings/WalletSettings.js:257]: @yakovkaravelov this is not good still.
If the current `delegationStakePoolStatus ` is DELEGATING and the `lastDelegationStakePoolStatus ` is UNDELEGATED, `isDelegating ` variable will be true and it should be false.

**Review by nikolaglumac — CHANGES_REQUESTED** [review]

**Review by nikolaglumac — APPROVED** [review]
Great work @yakovkaravelov 💯

**Comment by dbatiste** [comment]
My understanding is that this feature is done, but disabled behind a flag.  Any idea when this will be enabled?

**Comment by daniloprates** [comment]
> My understanding is that this feature is done, but disabled behind a flag. Any idea when this will be enabled?

@dbatiste please see: https://github.com/input-output-hk/daedalus/pull/2636

This feature is currently blocked by an issue in the backend. We are working to have this done as soon as we can.

**Pull request merged**
